### PR TITLE
Update to use the new dat-worker-pool

### PR DESCRIPTION
**Pull request by jcredding**
This updates dat-tcp to use the new dat-worker-pool.
dat-worker-pool's interface has changed but functionally it works
the same as it previously did.

When a server is initialized it now takes a worker class and a
set of options instead of just taking options. This is because
dat-worker-pool now requires a worker class when its initialized.
A block is no longer part of the server or dat-worker-pool's
interface. The logic for processing a work item (a client socket
for dat-tcp) is now part of the worker class that is passed in.
The server also no longer takes worker callbacks, these should be
added to the worker class thats passed in.

The server now also builds a worker pool when its initialized
instead of waiting till its started. This was previously made
possible when dat-worker-pool was changed to not start
automatically but dat-tcp didn't take advantage of it til now.
This avoids some conditionals and having to set/unset the worker
pool ivar as the server is started/stopped.

As part of this change, dat-tcp now also provides a `Worker` mixin
which for now only brings in dat-worker-pools `Worker` mixin. It
also requires the worker class that is passed to the server use
this mixin.

This also includes a number of small changes related to using the
new dat-worker-pool and the server interface changing:

* The server no longer automatically closes the client socket. This
  is alittle aggressive for something as generic as dat-tcp. Not
  all servers will always want to do this and it doesn't work
  nicely with the new worker class pattern of dat-worker-pool.
* The server no longer takes a min and max number of workers
  because dat-worker-pool no longer supports a min or max. It now
  defaults to 2 workers (its previous min value).
* Renamed the passed in client file descriptors. This confused me
  because the local variable and the method were the same name.
* Removed the thread status from the inspect. I'd rather not expose
  this level of detail about how the server works.
* Changed the work loop organization to try and make it more
  readable. I didn't find it very approachable when I was trying to
  understand its logic.
* Updated styles and conventions to our latest standards
  throughout.

@kellyredding - Ready for review.

### CODE REVIEW

**Comment by kellyredding** [comment]
@jcredding "dat-worker-pool *has* interface has changed..." in first paragraph of commit msg.

**Comment by kellyredding** [comment]
@jcredding "...part of the server or *dat-worker-pool's* interface..." in second paragraph of commit msg.

**Comment by kellyredding** [comment]
@jcredding "...take advantage of it *til* now." in third paragraph of commit msg.

File: lib/dat-tcp.rb [hunk 1]
```diff
@@ -4,37 +4,40 @@
...
 
       @logger = DatTCP::Logger.new(@debug)
 
-      @tcp_server       = nil
-      @work_loop_thread = nil
-      @worker_pool      = nil
-      @signal = Signal.new(:stop)
+      options[:num_workers] ||= DEFAULT_NUM_WORKERS
+      @worker_pool = DatWorkerPool.new(worker_class, {
+        :num_workers   => options[:num_workers],
```
Comment (kellyredding) [inline lib/dat-tcp.rb]: @jcredding wouldn't it be better to do `:num_workers   => options[:num_workers] || DEFAULT_NUM_WORKERS,` here and not mess with the passed in options hash on line 32?  Plus more consistent when option defaulting above.

File: lib/dat-tcp.rb [hunk 2]
```diff
@@ -95,103 +98,64 @@ def halt(wait = false)
...
+    def teardown
+      logger.info "Stopping work loop..."
+      unless @signal.halt?
+        logger.info "Shutting down worker pool"
+        @worker_pool.shutdown(@shutdown_timeout)
+      end
+      unless @signal.pause?
+        logger.info "Closing TCP server connection"
+        self.stop_listen
+      end
```
Comment (kellyredding) [inline lib/dat-tcp.rb]: @jcredding minor, and I just noticed b/c of this awesome set of changes, but don't you want to stop listening *before* shutting down the worker pool?

File: test/unit/dat-tcp_tests.rb [hunk 3]
```diff
@@ -218,17 +209,11 @@ class WorkLoopWithWorkTests < ListenAndRunningTests
...
     end
 
-    should "allow retrieving the client sockets file descriptors" do
-      assert_includes @client.fileno, subject.client_file_descriptors
-    end
-
-    should "client file descriptors should still be accessible after its paused" do
-      @io_select_stub.set_data_on_signal_pipe
-      @server.pause true
-      assert_includes @client.fileno, subject.client_file_descriptors
```
Comment (kellyredding) [inline test/unit/dat-tcp_tests.rb]: @jcredding why did you remove this test?

**Comment by kellyredding** [comment]
@jcredding so this is really awesome man - great stuff!  I *really* like the refactors and how this cleaned up the implementation.  Really really cool stuff here.

I made a few minor comments.  I'll look again once addressed.

File: lib/dat-tcp.rb (hunk 2 above)
Comment (jcredding) [inline lib/dat-tcp.rb]: @kellyredding - Yeah that's a good point, I switched it.

File: test/unit/dat-tcp_tests.rb (hunk 3 above)
Comment (jcredding) [inline test/unit/dat-tcp_tests.rb]: @kellyredding - I originally removed it because the client file descriptors are known no matter what the state of the server, but its not a bad test to have for any future changes so I restored it. Its a good check to ensure we still have access to the client file descriptors after we paused, otherwise Sanford can't hot restart correctly.

**Comment by jcredding** [comment]
@kellyredding - I've updated this.

**Comment by kellyredding** [comment]
@jcredding looks good :-1:
